lang/LangPrimSource/PyrUnixPrim.cpp

```diff
@@ -341,6 +342,80 @@ int prGMTime(struct VMGlobals *g, int numArgsPushed)
 	return errNone;
 }
 
+int prDateFromString(struct VMGlobals *g, int numArgsPushed)
+{
+	PyrSlot *dateObjectSlot = g->sp - 2;
+	PyrSlot *timeSlot = g->sp - 1;
+	PyrSlot *formatSlot = g->sp;
+
+	constexpr auto len = 1024;
+	char timeString[len];
+	if (slotStrVal(timeSlot, timeString, len)) return errWrongType;
+	char formatString[len];
+	if (slotStrVal(formatSlot, formatString, len)) return errWrongType;
+
+	using namespace boost::posix_time;
+	auto facet = new time_input_facet(); // will be owned/deleted by the locale
+	facet->format(formatString);
+	std::istringstream ss(timeString);
+	ss.imbue(std::locale(ss.getloc(), facet));
+	ptime pt;
+	if (!(ss >> pt)) {
+		error("time parsing failed\n");
+		return errFailed;
+	}
+	auto tm0 = to_tm(pt);
+	
+	tm0.tm_isdst = -1; // attempt to determine if Daylight Saving Time in effect
+	auto timePoint = std::chrono::system_clock::from_time_t(mktime(&tm0));
+	fillSlotsFromTime(dateObjectSlot, &tm0, timePoint);
+	return errNone;
+}
+
+// Given an incomplete Date structure (missing dayOfWeek and
+// rawSeconds), calculate and fill in those properties.
+int prDateResolve(struct VMGlobals *g, int numArgsPushed)
+{
+	PyrSlot *a = g->sp;
+	PyrSlot *slots = slotRawObject(a)->slots;
+	
+	if (IsNil(slots + 0)) {
+		SetNil(a);
+		return errNone;
+	}
+	if (!IsNil(slots+6) || !IsNil(slots+7)) {
+		error("dayOfWeek or rawSeconds are already defined\n");
+		return errFailed;
+	}
```

Comment:
i think i may have been unclear before.

why is this a failure condition? as opposed to filling in dayOfWeek if missing and rawSeconds if missing, and returning without error as long as type checks pass?

Reply:
Well, I'd put some logic in `Date.new` that only calls "resolve" if one of those two last arguments is Nil (see Date.sc).  To be "conservative" I put the above code in prDateResolve() that ensures we're being called under those (expected) conditions (one of them Nil).  But as you suggest, maybe this is too extreme.  I guess I'll just remove this check in the C++ code.  The code that follows will work correctly, filling those properties in (regardless of whether or not they were Nil, they will be overwritten -- the incoming values aren't saved).  But I think that's okay, and sounds like you do, too.  Thanks.

Reply:
To be clear, with the new code (where these checks are removed), if I do one of these:

```supercollider
Date(1977, 5, 25, 13, 30, 22, dayOfWeek: 2);
// -> Wed May 25 13:30:22 1977
```

Even if the incoming `dayOfWeek` (2) is passed in wrong, the new method will return a valid (corrected) date (`Wed May 25 13:30:22 1977`, where the valid dayOfWeek (Wednesday) is 3):

```supercollider
Date(1977, 5, 25, 13, 30, 22, rawSeconds: 1234567.8);
// -> Wed May 25 13:30:22 1977
```

Even if the `rawSeconds` is wrong here, a valid date will be returned (same as above).  This "valid" Date will have a computed `rawSeconds` of 233407822, not the 1234567.8 the user passed into new.  The other option in the case of passing in rawSeconds would be to ignore all the other arguments (year/month/day/etc) and give priority to the rawSeconds, and fill in all the rest...  That's easy to do too, if you think it's better.  Personally, I think if you specify all the other params, it's most likely you want them to "rule" (this is why I suggested it would be better if neither dayOfWeek nor rawSeconds were arguments to `new`)...